<font color='blue' size=5><b>Комментарий ревьюера</b></font>

<font color='blue'>Привет! Меня зовут Тимофей Вилков, и я буду проверять твой проект. Спасибо за проделанную работу! Предлагаю общаться на «ты».</font>

<font color='blue'>Я буду использовать различные цвета, чтобы было удобнее воспринимать мои комментарии:</font>



<div class="alert alert-block alert-success">
✔️ <b>Зеленый текст</b> - все отлично
</div>

<div class="alert alert-block alert-warning">
⚠️ <b>Желтый текст</b> - сделано все правильно, однако есть рекомендации, на что стоит обратить внимание
</div>

<div class="alert alert-block alert-danger">
❌ <b>Красный текст</b> - красный текст - есть недочеты
</div>

---

<font color='blue'>Пожалуйста, не удаляй мои комментарии в случае возврата работы, так будет проще разобраться, какие были недочеты, а также сразу увидеть исправленное. </font>

Ответы на мои комментарии лучше тоже помечать.
Например:

<div class="alert alert-block alert-info">
<b>Комментарий студента:</b> 
Текст комментария, вопроса, объяснения
</div>

<font color='blue'><b>Давай смотреть, что получилось!</b></font>

<div class="alert alert-block alert-success">
✔️ Привет ещё раз! Новые рекомендации я пометил вот так: Ревью № 2
</div>
</div>

# Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

# Описание столбцов

children — количество детей в семье

days_employed — общий трудовой стаж в днях

dob_years — возраст клиента в годах

education — уровень образования клиента

education_id — идентификатор уровня образования

family_status — семейное положение

family_status_id — идентификатор семейного положения

gender — пол клиента

income_type — тип занятости

debt — имел ли задолженность по возврату кредитов

total_income — ежемесячный доход

purpose — цель получения кредита

## Шаг 1. Откройте файл с данными и изучите общую информацию

In [55]:
import pandas as pd
data = pd.read_csv('/datasets/data.csv')
display(data.head(30))

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   
20         1   -1311.604166         36              среднее             1   
21         1    -253.685166         33              среднее             1   
22         1   -1766.644138         24              среднее             1   
23         0    -272.981385         21               высшее             0   
24         1  338551.952911         57              среднее             1   
25         0  363548.489348         67              среднее             1   
26         0            NaN         41              среднее             1   
27         0    -529.191635         28               высшее             0   
28         1    -717.274324         26               высшее             0   
29         0            NaN         63              среднее             1   

            family_status  family_status_id gender  income_type  debt  \
0         женат / замужем                 0      F    сотрудник     0   
1         женат / замужем                 0      F    сотрудник     0   
2         женат / замужем                 0      M    сотрудник     0   
3         женат / замужем                 0      M    сотрудник     0   
4        гражданский брак                 1      F    пенсионер     0   
5        гражданский брак                 1      M    компаньон     0   
6         женат / замужем                 0      F    компаньон     0   
7         женат / замужем                 0      M    сотрудник     0   
8        гражданский брак                 1      F    сотрудник     0   
9         женат / замужем                 0      M    сотрудник     0   
10        женат / замужем                 0      M    компаньон     0   
11        женат / замужем                 0      F    сотрудник     0   
12       гражданский брак                 1      M    пенсионер     0   
13        женат / замужем                 0      F    сотрудник     0   
14       гражданский брак                 1      F    компаньон     1   
15        женат / замужем                 0      F    сотрудник     0   
16        женат / замужем                 0      F    сотрудник     0   
17       гражданский брак                 1      M    сотрудник     0   
18         вдовец / вдова                 2      F    пенсионер     0   
19              в разводе                 3      F    сотрудник     0   
20        женат / замужем                 0      M    сотрудник     0   
21       г

###  Изучим общую информацию

In [56]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [57]:
data.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

**Вывод**
1) Пропуски замечены только в столбцах "days_employed" и "total_income". Возможно, заемщик не хотел указывать реальные свои данные, так как они ниже необходимомого минимума для одобрения кредита. Однако, вероятность получения кредита без указания этих данных выше, чем с ними. Так же, это может быть связано с типом занятости: пенсионер, госслужащий, студент. 

2) В столбце 'children' минимальное значние -1. Это может говорить о неверно внесенной или неверно подгруженной информации. Максимальное значение 20. Не смотря на то, что число уже не отрицательное, все равно наводит на мысль об ошибке в данных

3) Отрицательные значения по количеству отработанных дней 'days_employed' - возможно ли, что если взять модуль значений, то данные будут более пригодны для анализа. Так же такие значения наводят на мысль о технологической ошибке: при копировании данных, либо в процессе записи данных мог произойти сбой. 

4) Среднее значение по количеству отработанных дней 'days_employed' - 63047 дней (это почти 173 года). Больше похоже на человеческий фактор ошибки: при внесеннии данных заемщик или работних кредитного отдела (смотря кто именно вносил данные) мог совешить опечатку, отсюда нереалистичное среднее по данным.

5) Возраст по некоторым строкам 'dob_years' равен 0. Вероятней всего, снова играет роль человеческий фактор: заемщик не захотел указывать свой возраст в силу личных обстоятельств. Но технологические ошибки тоже не стоит опускать: возраст мог быть указан некорректно или в несоответсвующем формате и система сохранила значения в виде 0.



<div class="alert alert-block alert-success">
✔️Отличная работа с первичным выводом! Очень радует, что можешь извлечь так много информации из данных при первичном осмотре.
</div>

## Шаг 2. Предобработка данных

### Обработка пропусков

Проверяем на соответствие пропусков в столбцах 'days_employed' и 'total_income' 

In [58]:
data[(data['total_income'].isnull() == True) & (data['days_employed'].isnull() == True)].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2174 entries, 12 to 21510
Data columns (total 12 columns):
children            2174 non-null int64
days_employed       0 non-null float64
dob_years           2174 non-null int64
education           2174 non-null object
education_id        2174 non-null int64
family_status       2174 non-null object
family_status_id    2174 non-null int64
gender              2174 non-null object
income_type         2174 non-null object
debt                2174 non-null int64
total_income        0 non-null float64
purpose             2174 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 220.8+ KB


**Вывод**: В строках, где отутствуют данные по столбцам "total_income", так же отсутствуют данные по "days_employed".

In [59]:
# рассмотрим столбец income_type
data['income_type'].value_counts()

сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
предприниматель        2
безработный            2
в декрете              1
студент                1
Name: income_type, dtype: int64

In [60]:
#Найдем медианные значения дохода по каждому типу занятости

employee_median_total_income = data[data['income_type'] == 'сотрудник']['total_income'].median()
companion_median_total_income = data[data['income_type'] == 'компаньон']['total_income'].median()
retire_median_total_income = data[data['income_type'] == 'пенсионер']['total_income'].median()
governer_median_total_income = data[data['income_type'] == 'госслужащий']['total_income'].median()
business_median_total_income = data[data['income_type'] == 'предприниматель ']['total_income'].median()
student_median_total_income = data[data['income_type'] == 'студент']['total_income'].median()
decree_median_total_income = data[data['income_type'] == 'в декрете']['total_income'].median()

print('Медиана ежемесячного дохода сотрудников составляет', employee_median_total_income)
print('Медиана ежемесячного дохода компаньонов составляет', companion_median_total_income)
print('Медиана ежемесячного дохода пенсионеров составляет', retire_median_total_income)
print('Медиана ежемесячного дохода госслужащих составляет', governer_median_total_income)
print('Медиана ежемесячного дохода предпринимателей составляет', business_median_total_income)
print('Медиана ежемесячного дохода студентов составляет', student_median_total_income)
print('Медиана ежемесячного дохода в декрете составляет', decree_median_total_income)

Медиана ежемесячного дохода сотрудников составляет 142594.39684740017
Медиана ежемесячного дохода компаньонов составляет 172357.95096577113
Медиана ежемесячного дохода пенсионеров составляет 118514.48641164352
Медиана ежемесячного дохода госслужащих составляет 150447.9352830068
Медиана ежемесячного дохода предпринимателей составляет nan
Медиана ежемесячного дохода студентов составляет 98201.62531401133
Медиана ежемесячного дохода в декрете составляет 53829.13072905995


In [61]:
# Заполним пропущенные значения
data.loc[data['income_type'] == 'сотрудник', 'total_income']=data.loc[data['income_type'] == 'сотрудник', 'total_income'].fillna(value=employee_median_total_income)

data.loc[data['income_type'] == 'компаньон', 'total_income']=data.loc[data['income_type'] == 'компаньон', 'total_income'].fillna(value=companion_median_total_income)

data.loc[data['income_type'] == 'пенсионер', 'total_income']=data.loc[data['income_type'] == 'пенсионер', 'total_income'].fillna(value=retire_median_total_income)

data.loc[data['income_type'] == 'госслужащий', 'total_income']=data.loc[data['income_type'] == 'госслужащий', 'total_income'].fillna(value=governer_median_total_income)

data.loc[data['income_type'] == 'студент', 'total_income']=data.loc[data['income_type'] == 'студент', 'total_income'].fillna(value=student_median_total_income)

data.loc[data['income_type'] == 'в декрете', 'total_income']=data.loc[data['income_type'] == 'в декрете', 'total_income'].fillna(value=decree_median_total_income)

# Медианное значение дохода предпринимателей не посчитался, возможно по ним отсутствуют данные. Можно удалить эти строки
#так как их всего две и они не повлияют на значения результатов

display(data.head(50))



children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   
20         1   -1311.604166         36              среднее             1   
21         1    -253.685166         33              среднее             1   
22         1   -1766.644138         24              среднее             1   
23         0    -272.981385         21               высшее             0   
24         1  338551.952911         57              среднее             1   
25         0  363548.489348         67              среднее             1   
26         0            NaN         41              среднее             1   
27         0    -529.191635         28               высшее             0   
28         1    -717.274324         26               высшее             0   
29         0            NaN         63              среднее             1   
30         1  335581.668515         62              среднее             1   
31         0   -1682.083438         47            начальное             3   
32         0   -4649.910832         34              СРЕДНЕЕ             1   
33         0   -1548.637544         48              среднее             1   
34         0   -4488.067031         35              среднее             1   
35         0  394021.072184         68              среднее             1   
36         0    -176.216688         33              среднее             1   
37         0   -6448.810860         43               ВЫСШЕЕ             0   
38         0    -597.881827         25               высшее             0   
39         0    -650.587796         31               высшее             0   
40         0   -1030.362606         30              среднее             1   
41         0            NaN         50              среднее             1   
42         0   -1257.496190         20  неоконченное высшее             2   
43         0   -4375.681384         43  неоконченное высшее             2   
44         1   -1362.041728         26              СРЕДНЕЕ             1   
45         1   -1039.451741         49              среднее             1   
46         2   -2262.712304         37              среднее             1   
47         1   -2689.137274         33               высшее             0   
48         0   -3341.067886         45              среднее             1   
49         0   -1181.443228         54              среднее             1   

            family_status  family_status_id gender  income_type  debt  \
0  

In [62]:
#найдем ср. знач  стажа 
days_employed_mean = data['days_employed'].mean()

#заменим отрицательные числа на положительные в стаже
def negative_to_pozitive(value):
    if value < 0:
        value *= -1
        return value
    else:
        return value
data['days_employed'] = data['days_employed'].apply(negative_to_pozitive)
data['days_employed'] = data['days_employed'].fillna(days_employed_mean)



display(data.head(30))

children  days_employed  dob_years            education  education_id  \
0          1    8437.673028         42               высшее             0   
1          1    4024.803754         36              среднее             1   
2          0    5623.422610         33              Среднее             1   
3          3    4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0     926.185831         27               высшее             0   
6          0    2879.202052         43               высшее             0   
7          0     152.779569         50              СРЕДНЕЕ             1   
8          2    6929.865299         35               ВЫСШЕЕ             0   
9          0    2188.756445         41              среднее             1   
10         2    4171.483647         36               высшее             0   
11         0     792.701887         40              среднее             1   
12         0   63046.497661         65              среднее             1   
13         0    1846.641941         54  неоконченное высшее             2   
14         0    1844.956182         56               высшее             0   
15         1     972.364419         26              среднее             1   
16         0    1719.934226         35              среднее             1   
17         0    2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0   10038.818549         48              СРЕДНЕЕ             1   
20         1    1311.604166         36              среднее             1   
21         1     253.685166         33              среднее             1   
22         1    1766.644138         24              среднее             1   
23         0     272.981385         21               высшее             0   
24         1  338551.952911         57              среднее             1   
25         0  363548.489348         67              среднее             1   
26         0   63046.497661         41              среднее             1   
27         0     529.191635         28               высшее             0   
28         1     717.274324         26               высшее             0   
29         0   63046.497661         63              среднее             1   

            family_status  family_status_id gender  income_type  debt  \
0         женат / замужем                 0      F    сотрудник     0   
1         женат / замужем                 0      F    сотрудник     0   
2         женат / замужем                 0      M    сотрудник     0   
3         женат / замужем                 0      M    сотрудник     0   
4        гражданский брак                 1      F    пенсионер     0   
5        гражданский брак                 1      M    компаньон     0   
6         женат / замужем                 0      F    компаньон     0   
7         женат / замужем                 0      M    сотрудник     0   
8        гражданский брак                 1      F    сотрудник     0   
9         женат / замужем                 0      M    сотрудник     0   
10        женат / замужем                 0      M    компаньон     0   
11        женат / замужем                 0      F    сотрудник     0   
12       гражданский брак                 1      M    пенсионер     0   
13        женат / замужем                 0      F    сотрудник     0   
14       гражданский брак                 1      F    компаньон     1   
15        женат / замужем                 0      F    сотрудник     0   
16        женат / замужем                 0      F    сотрудник     0   
17       гражданский брак                 1      M    сотрудник     0   
18         вдовец / вдова                 2      F    пенсионер     0   
19              в разводе                 3      F    сотрудник     0   
20        женат / замужем                 0      M    сотрудник     0   
21       г

In [63]:
# Средний стаж безработных слишком большой, удалим строки
#Так же удалим строки про предпринимателей, так как по ним данные не посчитались и их отсутствие не критично

data = data[data['income_type'] != 'безработный']
data = data[data['income_type'] != 'предприниматель']
data['income_type'].value_counts()

сотрудник      11119
компаньон       5085
пенсионер       3856
госслужащий     1459
в декрете          1
студент            1
Name: income_type, dtype: int64

In [64]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21521 entries, 0 to 21524
Data columns (total 12 columns):
children            21521 non-null int64
days_employed       21521 non-null float64
dob_years           21521 non-null int64
education           21521 non-null object
education_id        21521 non-null int64
family_status       21521 non-null object
family_status_id    21521 non-null int64
gender              21521 non-null object
income_type         21521 non-null object
debt                21521 non-null int64
total_income        21521 non-null float64
purpose             21521 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.1+ MB


В стоблце "days_employed" заменили пропуски на среднее значение, убрали отрицательные значения (привели к модулю).

В столбце "total_income" пропуски заменили медианным значением по каждому типу занятости.

Строки с безработными заемщиками были удалены по причине слишком большого стажа, таких строк всего 2, и их удалением критично повлиять на результаты оценки различных факторов на погашение кредита не может. Так же удалили 2 строки по типу "предприниматели", так как по ним не посчиталась медиана, а значит данные никак кроме 0 не заполнить - следовательно удаление этих строк так же не повлияет на результат

<div class="alert alert-block alert-danger">
❌ Доход заявителя очень сильно отличается от его должности. Этой информации в явном виде нет, однако, можно косвенно это учесть через переменную "тип занятости". Так заполнить пропуски удастся гораздо точнее. Для этого можно сгруппировать заявителей по типу занятости и для каждой категории заполнить значения медианой. 

</div>

<div class="alert alert-block alert-info">
<b>Комментарий студента:</b> 
Перезаполнила данные, немного дополнила вывод, исправляла в коде, отдельно ничего не выносила
</div>

<div class="alert alert-block alert-success">
✔️ Ревью № 2: Отлично
</div>
</div>

In [65]:
#заменим кол-во детей "20" на "2"
data['children'] = data['children'].replace(20, 2)

# заменим количество детей с "-1" на "1"
data['children'] = data['children'].replace(-1, 1)

data['children'].value_counts()

0    14146
1     4864
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

Заменили отрицательные значения в графе "children", а так же поменяли значение количество 20 детей на 2, так как мы уже заметили ранее, что вероятней всего произошла ошибка при записи данных и значение должно было быть однозначное.

<div class="alert alert-block alert-success">
✔️Молодец, что нашла "брак" в данных и его устранил
</div>

In [66]:
# заменим значение возраста заемщика 0 на медиану значений возраста заемщиков:
data['dob_years'] = data['dob_years'].replace(0, data['dob_years'].median())
data['dob_years'].value_counts()

42    698
35    617
40    609
41    607
34    603
38    598
33    581
39    573
31    559
36    555
44    547
29    545
30    540
48    538
37    537
50    514
43    513
32    510
49    508
28    503
45    496
27    492
56    487
52    484
47    480
54    479
46    475
58    460
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
66    183
22    183
67    167
21    111
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

### Замена типа данных

In [67]:
# заменим столбы с типом данных float64 на int64 и проверим как произошла замена:
data['days_employed'] = data['days_employed'].astype('int')
data['total_income'] = data['total_income'].astype('int')
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21521 entries, 0 to 21524
Data columns (total 12 columns):
children            21521 non-null int64
days_employed       21521 non-null int64
dob_years           21521 non-null int64
education           21521 non-null object
education_id        21521 non-null int64
family_status       21521 non-null object
family_status_id    21521 non-null int64
gender              21521 non-null object
income_type         21521 non-null object
debt                21521 non-null int64
total_income        21521 non-null int64
purpose             21521 non-null object
dtypes: int64(7), object(5)
memory usage: 2.1+ MB


**Вывод**

Был выбран именно метод astype(), так как было необходимо перевести данные именно в целочисленные, для более удобного анализа. 
Метод to_numeric() нам бы подошел, если бы мы хотели перевести все количественные данные из целочисленных в данные типа float64.

Данные, которые представлены в строковых значених - нет необходимости переводить в количественные типы.

### Обработка дубликатов

Были замечены повторяющиеся значения в столбце "education"

In [68]:
# рассмотрим столбец education
data['education'].value_counts()

среднее                13749
высшее                  4716
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   267
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
УЧЕНАЯ СТЕПЕНЬ             1
Ученая степень             1
Name: education, dtype: int64

Значения повторяются, буквы разного регистра

In [69]:
#приведем все значния к нижнему регистру
data['education'] = data['education'].str.lower()

# проверка на наличие дубликатов
data['education'].unique()

array(['высшее', 'среднее', 'неоконченное высшее', 'начальное',
       'ученая степень'], dtype=object)

Таким образом мы избавились от повторяющихся значений, приведя их к единому стилю написания

<div class="alert alert-block alert-success">
✔️Здорово, теперь в датасете нет неявных дубликатов, это радует:)
</div>

In [70]:
#проверим на наличие дубликатов
data.duplicated().sum()

72

In [71]:
# удалим дубликаты 
data = data.drop_duplicates().reset_index()

#проверим снова на их наличие
data.duplicated().sum()

0

### Лемматизация

In [72]:
# рассмотрим столбец purpose
data['purpose'].value_counts()

свадьба                                   791
на проведение свадьбы                     766
сыграть свадьбу                           765
операции с недвижимостью                  675
покупка коммерческой недвижимости         661
операции с жильем                         652
операции с коммерческой недвижимостью     650
покупка жилья для сдачи                   650
жилье                                     646
покупка жилья                             646
покупка жилья для семьи                   638
строительство собственной недвижимости    635
недвижимость                              633
операции со своей недвижимостью           627
строительство жилой недвижимости          624
покупка недвижимости                      621
покупка своего жилья                      620
строительство недвижимости                619
ремонт жилью                              606
покупка жилой недвижимости                605
на покупку своего автомобиля              505
заняться высшим образованием      

In [73]:
# лемматизируем столбец "purpose"
# для этого определим функцию row_lemma(purpose)

from pymystem3 import Mystem
m = Mystem()

def row_lemma(purpose):
    lemma = ' '.join(m.lemmatize(purpose))
    return lemma

# создадим столбец purpose_lem, в котором будут леммы столбца purpose
data['purpose_lem'] = data['purpose'].apply(row_lemma)

# посчитаем количество уникальных лемм и выведем их в порядке убывания:
from collections import Counter
purpose_lem_count = Counter(data['purpose_lem'])
purpose_lem_count.most_common()

[('автомобиль \n', 972),
 ('свадьба \n', 791),
 ('на   проведение   свадьба \n', 766),
 ('сыграть   свадьба \n', 765),
 ('операция   с   недвижимость \n', 675),
 ('покупка   коммерческий   недвижимость \n', 661),
 ('операция   с   жилье \n', 652),
 ('операция   с   коммерческий   недвижимость \n', 650),
 ('покупка   жилье   для   сдача \n', 650),
 ('покупка   жилье \n', 646),
 ('жилье \n', 646),
 ('покупка   жилье   для   семья \n', 638),
 ('строительство   собственный   недвижимость \n', 635),
 ('недвижимость \n', 633),
 ('операция   со   свой   недвижимость \n', 627),
 ('строительство   жилой   недвижимость \n', 624),
 ('покупка   недвижимость \n', 621),
 ('покупка   свой   жилье \n', 620),
 ('строительство   недвижимость \n', 619),
 ('ремонт   жилье \n', 606),
 ('покупка   жилой   недвижимость \n', 605),
 ('на   покупка   свой   автомобиль \n', 505),
 ('заниматься   высокий   образование \n', 496),
 ('сделка   с   подержанный   автомобиль \n', 486),
 ('на   покупка   подержать   авт

**Вывод**

В данном случае лемматизация удобнее стемминга, так как заемщики использовали целые словосочетания для описания цели. В дальнейшем после лемматизации будет удобнее пользоваться отбором по столбцу "purpose", выбирая словарную форму слова для поиска.

### Категоризация данных

In [74]:
# создадим столбец purpose_category на основании данных лемматизации столбца purpose_lem
def purpose_category(purpose_lem):
    if 'автомобиль' in purpose_lem:
        return "сделки с автомобилем"
    if "образование" in purpose_lem:
        return "образование"
    if "свадьба" in purpose_lem:
        return "свадьба"
    if "недвижимость" in purpose_lem or "строительство" in purpose_lem or "жилье" in purpose_lem:
        return "сделки с недвижимостью"
   
data['purpose_category'] = data['purpose_lem'].apply(purpose_category)

data['purpose_category'].value_counts()

сделки с недвижимостью    10808
сделки с автомобилем       4306
образование                4013
свадьба                    2322
Name: purpose_category, dtype: int64

<div class="alert alert-block alert-warning">
⚠️ Если хочешь чтобы условие в <code>if</code> выполнялось в случае выполнения одного из логических условий удобнее использовать <code>or</code>:
<code>
if "недвижимость" in purpose_lem or "строительство" in purpose_lem or "жилье" in purpose_lem:
    return "сделки с недвижимостью"
</code>

</div>

<div class="alert alert-block alert-info">
<b>Комментарий студента:</b> 
Подправила условие. Ты прав, так гораздо лучше
</div>

In [75]:
# удалим столбыцы purpose и purpose_lem
data.drop('purpose', axis=1, inplace=True) 
data.drop('purpose_lem', axis=1, inplace=True)

# проверим, как теперь выглядит таблица
data

index  children  days_employed  dob_years education  education_id  \
0          0         1           8437         42    высшее             0   
1          1         1           4024         36   среднее             1   
2          2         0           5623         33   среднее             1   
3          3         3           4124         32   среднее             1   
4          4         0         340266         53   среднее             1   
...      ...       ...            ...        ...       ...           ...   
21444  21520         1           4529         43   среднее             1   
21445  21521         0         343937         67   среднее             1   
21446  21522         1           2113         38   среднее             1   
21447  21523         3           3112         38   среднее             1   
21448  21524         2           1984         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21444  гражданский брак                 1      F   компаньон     0   
21445   женат / замужем                 0      F   пенсионер     0   
21446  гражданский брак                 1      M   сотрудник     1   
21447   женат / замужем                 0      M   сотрудник     1   
21448   женат / замужем                 0      F   сотрудник     0   

       total_income        purpose_category  
0            253875  сделки с недвижимостью  
1            112080    сделки с автомобилем  
2            145885  сделки с недвижимостью  
3            267628             образование  
4            158616                 свадьба  
...             ...                     ...  
21444        224791  сделки с недвижимостью  
21445        155999    сделки с автомобилем  
21446         89672  сделки с недвижимостью  
21447        244093    сделки с автомобилем  
21448         82047    сделки с автомобилем  

[21449 rows x 13 columns]

In [76]:
#разобьем на квантили наш столбец total_income
pd.qcut(data['total_income'], q=4)

0        (195799.0, 2265604.0]
1         (107620.0, 142594.0]
2         (142594.0, 195799.0]
3        (195799.0, 2265604.0]
4         (142594.0, 195799.0]
                 ...          
21444    (195799.0, 2265604.0]
21445     (142594.0, 195799.0]
21446    (20666.999, 107620.0]
21447    (195799.0, 2265604.0]
21448    (20666.999, 107620.0]
Name: total_income, Length: 21449, dtype: category
Categories (4, interval[float64]): [(20666.999, 107620.0] < (107620.0, 142594.0] < (142594.0, 195799.0] < (195799.0, 2265604.0]]

In [77]:
#Впишем в функцию полученные интервалы
#Первый квантиль будем считать 'Доходы ниже среднего'
#Второй квантиль будем считать 'Средние доходы'
#Третий квантиль будем считать 'Доходы выше среднего'
#Четвертый квантиль будем считать 'Высокие доходы'


def group_income(row):
    
    income = row['total_income']

    if income <= 107620:
        return 'Доходы ниже среднего'
   
    if 107620 < income <= 142594:
        return 'Средние доходы'
    
    if 142594 < income <= 195799:
        return 'Доходы выше среднего'
    
    if income > 195799:
        return 'Высокие доходы'

data['group_in'] = data.apply(group_income, axis=1)
data
data['group_in'].value_counts()


Средние доходы          5478
Доходы ниже среднего    5363
Высокие доходы          5362
Доходы выше среднего    5246
Name: group_in, dtype: int64

<div class="alert alert-block alert-danger">
❌ Есть несколько способов разбить переменную на бины лучше. Я бы разбил по квантилям — это можно сделать с помощью <code>pd.qcut()</code> , такой подход лучше экспертной оценки интервалов
<br>
Почитать можно в документации: 
<a href="https://pandas.pydata.org/docs/reference/api/pandas.qcut.html ">qcut</a>

</div>

<div class="alert alert-block alert-info">
<b>Комментарий студента:</b> 
Заменила способ разделения по доходам, но теперь получаются другие значения, которые практически равны друг другу, существенных различий нет
</div>

<div class="alert alert-block alert-success">
✔️ Ревью № 2: Можно использовать просто больше q. Например, 8
</div>
</div>

**Вывод**

После проведения категоризации, мы привели цели кредитования к четырем типам: сделки с автомобилем, сделки с недвижимостью, свадьба, образование. 

Так же разбили доходы на категории, для представления общей картины.

Это было сделано с целью дальнейшей оптимизации работы с данными: улучшение визуальных данных, уменьшение объема файла, удобная фильтрация по типу данных.

## Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [78]:
#Посчитаем процент заемщиков с задолженностью по возврату кредита

debt_per = data['debt'][(data['debt'] == 1)].count() / data['debt'].count()
print('{:.2%}'.format(debt_per), 'среди заемщиков не возвращали кредит вовремя')

8.11% среди заемщиков не возвращали кредит вовремя


In [79]:
# Построим сводную таблицу на основании исходной обработанной используя столбцы "children", "debt" и "family_status"

data_pivot_children = data.pivot_table(index=['children'], columns='debt', values = 'family_status', aggfunc='count')

# Для удобства обработки информации переименуем слобцы
names_columns = ['Добросовестный плательщик', 'Должник']
data_pivot_children.set_axis(names_columns, axis = "columns", inplace = True)

# Добавим еще столбцы с расчетом доли должников с разным количеством детей и столбец с отклонением от общей доли должников

data_pivot_children['Доля должников'] = data_pivot_children['Должник'] / data_pivot_children['Добросовестный плательщик']
data_pivot_children['Отклонение от общего уровня должников'] = data_pivot_children['Доля должников'] - debt_per
display(data_pivot_children)

Добросовестный плательщик  Должник  Доля должников  \
children                                                       
0                           13024.0   1063.0        0.081619   
1                            4410.0    444.0        0.100680   
2                            1926.0    202.0        0.104881   
3                             303.0     27.0        0.089109   
4                              37.0      4.0        0.108108   
5                               9.0      NaN             NaN   

          Отклонение от общего уровня должников  
children                                         
0                                      0.000496  
1                                      0.019558  
2                                      0.023758  
3                                      0.007986  
4                                      0.026985  
5                                           NaN

**Вывод**

На основании сводной таблицы наблюдается, что задолженности по кредитам увеличивается прямопропорционально появлению в семье детей.

<div class="alert alert-block alert-warning">
⚠️ Риски на семьях с 3 детьми ниже, чем риски у семей с двумя детьми. Поэтому мы не можем утверждать, что рост количества детей увеличивает риски. Тем не менее, действительно, заявители с детьми в среднем хуже по рискам. 

</div>

<div class="alert alert-block alert-danger">
❌ Не хватает твоих выводов, прокомментируй, пожалуйста, результаты аналитики. Какая по твоему мнению природа связи которую мы наблюдаем?

</div>

<div class="alert alert-block alert-info">
<b>Вывод:</b> 
    При появлении в семье последующих детей выростают расходы непосредственно на ребенка, многие отодвигают заботы о своевремнных выплатах на задний план, ставя в первую очередь благосостояние своих детей. Иными словами все больше погружаются в воспитание детей и становление их на ноги.Когда появлется третий ребенок, возможно к этому моменту родители уже научились распределять доходы таким образом, что если не вовремя платить по кредиту, то хотя бы избежать слишком большой просрочки платежа. Когда нет детей, то все средства уходят на собственное благосостояние и своевременная выплата кредита в интересах потребителя, чтобы не испортить себе кредитную историю.
</div>

<div class="alert alert-block alert-success">
✔️ Ревью № 2: Отличная гипотеза
</div>
</div>

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [80]:
# Построим сводную таблицу по примеру обработки предыдущего вопроса, используя столбцы "family_status", "debt" и "family_status_id"

data_pivot_family_status = data.pivot_table(index=['family_status'], columns='debt', values = 'family_status_id', aggfunc='count')


names_columns = ['Добросовестный плательщик', 'Должник']
data_pivot_family_status.set_axis(names_columns, axis = "columns", inplace = True)

data_pivot_family_status['Доля должников'] = data_pivot_family_status['Должник'] / data_pivot_family_status['Добросовестный плательщик']
data_pivot_family_status['Отклонение от общего уровня должников'] = data_pivot_family_status['Доля должников'] - debt_per
display(data_pivot_family_status)

Добросовестный плательщик  Должник  Доля должников  \
family_status                                                               
Не женат / не замужем                       2536      274        0.108044   
в разводе                                   1110       85        0.076577   
вдовец / вдова                               896       63        0.070312   
гражданский брак                            3760      388        0.103191   
женат / замужем                            11407      930        0.081529   

                       Отклонение от общего уровня должников  
family_status                                                 
Не женат / не замужем                               0.026922  
в разводе                                          -0.004546  
вдовец / вдова                                     -0.010810  
гражданский брак                                    0.022069  
женат / замужем                                     0.000406

**Вывод**

Среди заемщиков высокая доля должников среди женатых/замужних; вдовцы и в разводе - самые ответственные заемщики и чаще возвращают кредиты в срок

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

<div class="alert alert-block alert-danger">
❌ Здесь тоже было бы здорово увидеть твои гипотезы на этот счёт
</div>

<div class="alert alert-block alert-info">
<b>Вывод:</b> 
 У заемщиков, состоящих в браке, гораздо больше потребностей (Приобретение недвижимости, автомобиля, свадьба, обучать детей), соответственно больше поводов взять кредиты. Бывает так, что на семью более 2х, а то и 3х кредитов, отсюда и просрочки, что неправильно изначально расчитали свои возможности по выплатам и поэтому образуются задолженности. У разведенных или же вдовцов дела обстоят лучше с своевременным погашением, по причине сосредоточенности на себе и более четком контроле своей финансовой жизни. 

<div class="alert alert-block alert-success">
✔️Ревью № 2: Вполне возможно
</div>
</div>

In [81]:
#Так же построим сводную таблицу используя столбцы "group_in", "debt" и "family_status_id"
data_pivot_group_in = data.pivot_table(index=['group_in'], columns='debt', values='family_status_id', aggfunc='count')

name_columns = ['Добросовестный плательщик', 'Должник']
data_pivot_group_in.set_axis(names_columns, axis = "columns", inplace = True)

data_pivot_group_in['Доля должников'] = data_pivot_group_in['Должник'] / data_pivot_group_in['Добросовестный плательщик']
data_pivot_group_in['Отклонение от общего уровня должников'] = data_pivot_group_in['Доля должников'] - debt_per
display(data_pivot_group_in)

Добросовестный плательщик  Должник  Доля должников  \
group_in                                                                   
Высокие доходы                             4979      383        0.076923   
Доходы выше среднего                       4798      448        0.093372   
Доходы ниже среднего                       4937      426        0.086287   
Средние доходы                             4995      483        0.096697   

                      Отклонение от общего уровня должников  
group_in                                                     
Высокие доходы                                    -0.004200  
Доходы выше среднего                               0.012250  
Доходы ниже среднего                               0.005165  
Средние доходы                                     0.015574

**Вывод**

С Высокими доходами и Доходами выше среднего - самые ответственные заемщики, в то время как заемщики Средними доходами имеют тенденцию к задержке возратов кредита в срок.

<div class="alert alert-block alert-danger">
❌С чем это может быть связано, как считаешь?
</div>

<div class="alert alert-block alert-info">
<b>Вывод:</b> 
С первой частью все понятно: чем больше доход, тем проще и быстрее выплачивать кредит. Почему потребители с доходами выше срденего берут кредиты? Чаще это вопрос менталитета и комфортно - удобнее платить по частям, чем сразу расставаться с крупной суммой. Теперь к вопросу, почему со Средним доходом все таки больше должников: в этом случае играет роль финансовая стабильность. Потребитель попросту может не расчитать свои силы по размеру выплаты, считая изначально, что кредит на ту или иную сумму ему под силу. Так же заемщики из данной категории - это наемные работники, временные интервалы получения дохода от них не зависят , а следовательно дата платежа может быть назначена на день после получения аванса - образуется просрочка.

<div class="alert alert-block alert-success">
✔️ Ревью № 2: С тобой согласен, но ещё можно добавить, что заявителям с низким доходом тяжелее получить кредит, банк их проверяет тщательнее
</div>
</div>

- Как разные цели кредита влияют на его возврат в срок?

In [82]:
#Так же построим сводную таблицу используя столбцы "purpose_category", "debt" и "family_status_id"
data_pivot_purpose_category = data.pivot_table(index=['purpose_category'], columns='debt', values='family_status_id', aggfunc='count')

name_columns = ['Добросовестный плательщик', 'Должник']
data_pivot_purpose_category.set_axis(names_columns, axis = "columns", inplace = True)

data_pivot_purpose_category['Доля должников'] = data_pivot_purpose_category['Должник'] / data_pivot_purpose_category['Добросовестный плательщик']
data_pivot_purpose_category['Отклонение от общего уровня должников'] = data_pivot_purpose_category['Доля должников'] - debt_per
display(data_pivot_purpose_category)

Добросовестный плательщик  Должник  Доля должников  \
purpose_category                                                             
образование                                  3643      370        0.101565   
свадьба                                      2136      186        0.087079   
сделки с автомобилем                         3903      403        0.103254   
сделки с недвижимостью                      10027      781        0.077890   

                        Отклонение от общего уровня должников  
purpose_category                                               
образование                                          0.020442  
свадьба                                              0.005956  
сделки с автомобилем                                 0.022131  
сделки с недвижимостью                              -0.003233

**Вывод**

Заемщики с кредитом на автомобиль имеют более высокие риски по выплате в срок, в то время как заемщики с кредитом на недвижимость являются самыми ответственными.

<div class="alert alert-block alert-danger">
❌ Здесь тоже, прокомментируй, пожалуйста
</div>

<div class="alert alert-block alert-info">
<b>Вывод:</b> 
В данном случае мы видим, что должников с недвижимостью больше, но и самих сделок больше по отношению к остальным. Поэтому обратимся к Доли должников. К сделкам с недвижимостью потребители подходят гораздо ответственней, чем к остальному: если банк отберет машину за долги, это плохо, но не критично, а вот если банк за долги будет отбирать "крышу над головой" - это уже неприятная ситуация, которую все стремятся ни в коем случае не допустить. Так что скорее в случае с целями кредита на их выплату скорее влияет человеческий фактор, чем финансовый.

<div class="alert alert-block alert-success">
✔️ Ревью № 2: Здорово
</div>
</div>

## Шаг 4. Общий вывод

Для работы над проектом по исследованию надежности заемщиков были использованы:
* Python;
* Pandas;
* предобработка данных;
* лемматизация средствами PyMystem3.
Было проведено исследование влияния семейного положения заемщика и количества его детей на факт возврата кредита в срок на основании статистики о платежеспособности клиентов за предыдуший период.
В ходе исследования определены и заполнены пропуски, заменены типы данных на соответствующие хранящимся данным, проверено наличие дубликатов, а так же их ликвидация без ущерба для данных. Затем были выделены леммы в значениях столбца с целями кредита, данные были категоризированы по целям, а в последствии выявлены зависимости возврата кредита в срок от семейного положения и количества детей.
В 8.11% случаев кредит не возвращался вовремя.
Заемщики, не имеющие детей более надежны в отношении погашения платежа по кредиту в срок по сравнению с общим количеством должников. С появлением первого и второго ребенка риск просрочки погашения платежей увеличивается. Многодетные семьи являются наименее надежными в плане своевременного погашения кредитов, по отношению к общему уровню должников.
Надежными плательщиками кредитов являются вдовцы и вдовы, а так же разведенные; женатые и замужние заемщики состовляют почти половину от всех должников
Заемщики со средними доходами чаще остальных имеют задолженность по кредитам, а люди с доходами выше среднего выплачивают займы стабильнее.

<font color='blue'><b>Итоговый комментарий ревьюера</b></font>\
<font color='green'>Екатерина, хороший проект, видно, что ты можешь много показать. Почти все сделала правильно, нужно исправить лишь несколько деталей:) 
   

<font color='blue'>Что можно сделать лучше:</font>
<ul><font color='red'>1) Использовать больше бинов для переменной дохода </font></ul>
<ul><font color='red'>2) Добавить гипотезы о причинах связи между исследуемыми переменными</font></ul>

<font color='blue'><b>Спасибо, жду твоих исправлений и комментариев!</b></font>

<div class="alert alert-block alert-info">
<b>Вывод:</b> 
Ответила в синих рамках. Не разобралась как делать переносы, чтобы текст был более удобным для чтения. Буду очень благодарна, если подскажешь как сделать перенос и остаться в цветной рамке.

In [ ]:
#Перенос осуществляется с помощью тега <br>

<div class="alert alert-block alert-success">
✔️ Ревью №2: Екатерина, здорово, что устранила все недочёты, получился отличный проект. С удовольствием перевожу его в состояние "зачтено". Желаю дальнейших успехов в освоении курса.
</div>